docs/EXTENSIBILITY.md

```diff
@@ -0,0 +1,82 @@
+# Spice.ai Extensibility
+
+This document is an overview of all the interfaces and extension points in Spice.ai.
+
+## DataConnector
+
+A `DataConnector` is a component that represents the source of data to the Spice.ai runtime. Datasets specify which `DataConnector` to use via the `from` field in the dataset configuration. A `DataConnector` knows how to retrieve data, stream data updates, and writing data back to the source.
+
+```rust
+/// A `DataConnector` knows how to retrieve and modify data for a given dataset.
+///
+/// Implementing `get_all_data` is required, but `stream_data_updates` & `supports_data_streaming` is optional.
+/// If `stream_data_updates` is not supported for a dataset, the runtime will fall back to polling `get_all_data` and returning a
+/// `DataUpdate` that is constructed like:
+///
+/// ```rust
+/// DataUpdate {
+///    data: get_all_data(dataset),
+///    update_type: UpdateType::Overwrite,
+/// }
+/// ```
+pub trait DataConnector: Send + Sync {
```

Comment:
I'm not a fan of documenting code in a separate location from the code. It gets hard to remember to update both locations and they end up diverging. How about we list the interfaces and extensions here with a quick summary of what they are and then link to the code itself where we full document?

Reply:
I somewhat agree. I would use table format with links.